src/lib.rs

```diff
@@ -413,27 +388,29 @@ fn format_file(file: &Path, config: &Config) -> FileMap {
                                                 codemap.clone());
     let mut parse_session = ParseSess::with_span_handler(tty_handler, codemap.clone());
 
-    let krate = parse::parse_crate_from_file(file, Vec::new(), &parse_session).unwrap();
+    let main_file = match input {
+        Input::File(ref file) => file.clone(),
+        Input::Text(..) => PathBuf::from("stdin"),
+    };
+
+    let krate = match parse_input(input, &parse_session) {
+        Ok(krate) => krate,
+        Err(mut diagnostic) => {
+            diagnostic.emit();
+            panic!("Unrecoverable parse error");
+        }
+    };
```

Comment:
Returning `Result` here won't be super straightforward because diagnostic lifetime is bound to `parse_session` which is a local variable. 

I plan to tackle this when implementing non-zero exit status. 

And actually I don't think that returning `Result` will be a proper way to handle errors here (I'm not sure about this though). I think that `rustfmt` should be a robust tool, and that it should try to format as much as possible in spite of errors. So it seems not practical to return `Result`s with one error, because there may be several errors present. Instead I propose to report errors right away, continue execution if possible and return a flag that there were some errors. In this particular case I'd `emit` a diagnostic and return an empty `FileMap` and `FormatReport`.